.travis.yml

```diff
@@ -0,0 +1,47 @@
+dist: bionic
+language: cpp
+
+jobs:
+  include:
+  - os: linux
+    compiler: gcc
+    dist: bionic
+    addons:
+      apt:
+        packages:
+        - ccache
+        - cmake
+        - clang-format
+        - libgmp3-dev
+        - parallel
+    before_install:
+      - .travis/preinstall.sh
+    script:
+      - cd $TRAVIS_BUILD_DIR && mkdir build && cd build && cmake .. && make
+  - os: linux
+    compiler: clang
+    dist: bionic
+    addons:
+      apt:
+        packages:
+        - ccache
+        - cmake
+        - clang-format
+        - libgmp3-dev
+        - parallel
+    before_install:
+      - .travis/preinstall.sh
+    script:
+      - cd $TRAVIS_BUILD_DIR && mkdir build && cd build && cmake .. && make
```

Comment:
Can we add `make test` (or change to `make all test`) too? It takes ~3min on my laptop, and would be a nice smoke test.

Reply:
I would also support this, though I notice @no2chem mentioned earlier:
> In currently does not run tests or clang-format. It should in the future (or I could amend this PR to).
I am not entirely clear as to whether @no2chem had a particular reason for deferring the addition of `make test`.

Reply:
Yep, I'll add that in now.

Reply:
It seems there are some issues running the tests in Travis, in particular pyclient_tests and skvbc_tests fails on Linux and macos, and on macos TestThresholdBls fails. 

I think for now it would be best to get this in as soon as possible given all the -Werror issues, and then work on trying to get those tests working on Travis, what do you think @beerriot @upshaw-alex

Reply:
Those tests fail on linux because they are missing trio, which can be installed like so:

```
sudo apt install python3 python3-pip
python3 -m pip install --upgrade trio
```

The README mentions it under the optional python client section. It should be added as a strict requirement above though, since the python client is not optional for tests, and other parts of the tests also require trio.

My preference would be to get tests running on linux, and temporarily remove support for MacOS. It's not critical for us right now to be able to build on MacOS and it's eating more time than I'd like worrying about it at the moment. When builds and tests pass on MacOS we can add that support back in to travis.

Reply:
Well ironically, the build seems to work just fine in macos now (and not any other os, at least with -Werror), but you are right, the tests are another issue. Let me see what adding trio does.

Reply:
Thanks Michael. It may have something to do with the reintroduction of signal handlers in #126. That's the only thing I can think of that would have caused this, but I suppose it could be a deadlock elsewhere in test changes. A lot of code was recently merged in, but these tests passed for me just yesterday.

Reply:
I also really appreciate the work your doing to get CI going. Hopefully we'll have less of these frustrations soon.

Reply:
No worries --- thanks for being patient with my changes as well as the pointers to help me get aligned with the code base. I hope with CI more issues are caught early on.

Reply:
I just ran skvbc tests vebosely via ctest and see that asserts are firing. My guess is that asserts fire and then the signal handler hangs because shutdown still doesn't work properly. 

```
andrewstone@ubuntu:~/concord-bft/build$ ctest -R skvbc --verbose
UpdateCTestConfiguration  from :/home/andrewstone/concord-bft/build/DartConfiguration.tcl
Parse Config file:/home/andrewstone/concord-bft/build/DartConfiguration.tcl
UpdateCTestConfiguration  from :/home/andrewstone/concord-bft/build/DartConfiguration.tcl
Parse Config file:/home/andrewstone/concord-bft/build/DartConfiguration.tcl
Test project /home/andrewstone/concord-bft/build
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 15
    Start 15: skvbc_tests

15: Test command: /usr/bin/python3 "-m" "unittest" "test_skvbc" "test_skvbc_chaos" "test_skvbc_linearizability"
15: Test timeout computed to be: 1500
15: INFO 2019-07-25 12:43:18.685 (comm-factory)Using PlainUDP: IP=0.0.0.0, Port=3710
15: INFO 2019-07-25 12:43:18.700 (comm-factory)Using PlainUDP: IP=0.0.0.0, Port=3712
15: INFO 2019-07-25 12:43:18.709 (comm-factory)Using PlainUDP: IP=0.0.0.0, Port=3714
15: INFO 2019-07-25 12:43:18.715 (comm-factory)Using PlainUDP: IP=0.0.0.0, Port=3716
15: INFO 2019-07-25 12:43:18.725 (state-transfer)Creating BCStateTran object: myId_=1 fVal_=1 sizeOfReservedPage_=4096
15: INFO 2019-07-25 12:43:18.726 (state-transfer)Creating BCStateTran object: myId_=0 fVal_=1 sizeOfReservedPage_=4096
15: skvbc_replica: /home/andrewstone/concord-bft/bftengine/src/bftengine/BFTEngine.cpp:123: static bftEngine::Replica *bftEngine::Replica::createNewReplica(bftEngine::ReplicaConfig *, bftEngine::RequestsHandler *, bftEngine::IStateTransfer *, bftEngine::ICommunication *, bftEngine::MetadataStorage *): Assertion `false' failed.
15: skvbc_replica: /home/andrewstone/concord-bft/bftengine/src/bftengine/BFTEngine.cpp:123: static bftEngine::Replica *bftEngine::Replica::createNewReplica(bftEngine::ReplicaConfig *, bftEngine::RequestsHandler *, bftEngine::IStateTransfer *, bftEngine::ICommunication *, bftEngine::MetadataStorage *): Assertion `false' failed.
```

Reply:
It's almost like we should run these automated tests before allowing commits to go in :)

Reply:
I can take a look at this a bit later. I'm working on something else right now...

Reply:
@andrewjstone Awesome. btw, should we run tests in verbose mode so we can debug their failures? Or will that result in too much output do you think?

Reply:
Yes, almost certainly they should be run verbosely. It shouldn't be too much output. I think it's likely the only way we'll figure out what is going on.